modules/internal/CString.chpl

```diff
@@ -0,0 +1,440 @@
+  inline proc =(ref a: c_string_copy, b: c_string_copy) {
+    chpl_free_c_string_copy(a);
+    __primitive("=", a, b);
+  }
+
+  // A c_string_copy can always be used as a c_string.
+  inline proc _cast(type t, x: c_string_copy) where t == c_string {
+    return __primitive("cast", t, x);
+  }
+
+  extern proc chpl_bool_to_c_string(x:bool) : c_string;
+  inline proc _cast(type t, x: bool(?w)) where t == c_string {
+    return chpl_bool_to_c_string(x:bool);
+  }
+  inline proc _cast(type t, x: bool(?w)) where t == c_string_copy {
+    return __primitive("string_copy", chpl_bool_to_c_string(x:bool));
+  }
+
+  inline proc _cast(type t, x:enumerated) where t == c_string_copy {
+    // Use the compiler-generated enum to c_string conversion.
+    var cs = _cast(c_string, x);
+    return __primitive("string_copy", cs);
+  }
+
+  inline proc _cast(type t, x:integral) where t == c_string_copy {
+    extern proc integral_to_c_string_copy(x:int(64), size:uint(32), isSigned: bool) : c_string_copy ;
+    return integral_to_c_string_copy(x:int(64), numBytes(x.type), isIntType(x.type));
+  }
+
+  //
+  // casts from c_string to bool types
+  //
+  inline proc _cast(type t, x:c_string) where isBoolType(t)
+  {
+    pragma "insert line file info"
+    extern proc c_string_to_chpl_bool(x:c_string) : bool;
+    return c_string_to_chpl_bool(x) : t;
+  }
+
+  //
+  // casts from c_string to integer types
+  //
+  pragma "insert line file info"
+  extern proc c_string_to_int8_t  (x:c_string) : int(8);
+  pragma "insert line file info"
+  extern proc c_string_to_int16_t (x:c_string) : int(16);
+  pragma "insert line file info"
+  extern proc c_string_to_int32_t (x:c_string) : int(32);
+  pragma "insert line file info"
+  extern proc c_string_to_int64_t (x:c_string) : int(64);
+  pragma "insert line file info"
+  extern proc c_string_to_uint8_t (x:c_string) : uint(8);
+  pragma "insert line file info"
+  extern proc c_string_to_uint16_t(x:c_string) : uint(16);
+  pragma "insert line file info"
+  extern proc c_string_to_uint32_t(x:c_string) : uint(32);
+  pragma "insert line file info"
+  extern proc c_string_to_uint64_t(x:c_string) : uint(64);
+
+  inline proc _cast(type t, x:c_string) where t == int(8)
+    return c_string_to_int8_t(x);
+  inline proc _cast(type t, x:c_string) where t == int(16)
+    return c_string_to_int16_t(x);
+  inline proc _cast(type t, x:c_string) where t == int(32)
+    return c_string_to_int32_t(x);
+  inline proc _cast(type t, x:c_string) where t == int(64)
+    return c_string_to_int64_t(x);
+  inline proc _cast(type t, x:c_string) where t == uint(8)
+    return c_string_to_uint8_t(x);
+  inline proc _cast(type t, x:c_string) where t == uint(16)
+    return c_string_to_uint16_t(x);
+  inline proc _cast(type t, x:c_string) where t == uint(32)
+    return c_string_to_uint32_t(x);
+  inline proc _cast(type t, x:c_string) where t == uint(64)
+    return c_string_to_uint64_t(x);
```

Comment:
I don't think this is new (you just moved it) but I've never really been happy with the way that these work... Why can we cast from a c_string to 1? Why not only support that for Chapel strings? Why doesn't it use other I/O code (like an 'unformat' call e.g.)?